src/azul/plugins/metadata/anvil/__init__.py

```diff
@@ -333,11 +341,67 @@ def recurse(mapping: MetadataPlugin._FieldMapping, path: FieldPath):
         result[('contents', 'files')]['file_url'] = 'files.azul_file_url'
         return result
 
+    primary_keys_by_table = {
+        table['name']: one(table['primaryKey'])
+        for table in anvil_schema['tables']
+    }
+
+    foreign_keys_by_table = {
+        table['name']: [
+            (r['to']['table'], r['from']['column'])
+            for r in anvil_schema['relationships']
+            if r['from']['table'] == table['name']
+        ]
+        for table in anvil_schema['tables']
+    }
+
     def verbatim_pfb_entity_id(self, replica: JSON) -> str:
-        if replica['replica_type'] == 'duos_dataset_registration':
-            return replica['contents']['duos_id']
+        replica_type = replica['replica_type']
+        try:
+            primary_key = self.primary_keys_by_table[replica_type]
+        except KeyError:
+            if replica_type == 'duos_dataset_registration':
+                return replica['contents']['duos_id']
+            else:
+                return super().verbatim_pfb_entity_id(replica)
         else:
-            return super().verbatim_pfb_entity_id(replica)
+            return replica['contents'][primary_key]
+
+    def verbatim_pfb_relations(self,
+                               replica: JSON
+                               ) -> list[tuple[str, str]]:
+        table_name, contents = replica['replica_type'], replica['contents']
+        try:
+            foreign_keys = self.foreign_keys_by_table[table_name]
+        except KeyError:
+            if table_name == 'duos_dataset_registration':
+                return [('anvil_dataset', contents['dataset_id'])]
+            else:
+                return super().verbatim_pfb_relations(replica)
+        else:
+            return [
+                (foreign_table_name, foreign_key)
+                for (foreign_table_name, foreign_key_column) in foreign_keys
+                for foreign_key in always_iterable(contents[foreign_key_column])
```

Comment:
Using `always_iterable` goes against our principle of strictness because it hides which of the three cases (iterable, non-iterable and None) are expected and which are not. If all three cases *are* expected, a comment should state that, as it would be surprising. There is an existing use of `always_iterable` to which these consideration should also be applied, in a separate commit.

Reply:
Agreed at PL to briefly explain the three cases in a comment for both uses.